[3.x] Reactive messaging (pulsar) native integration test fails after #54794

### Describe the bug

After merging https://github.com/quarkusio/quarkus/pull/54794 into the 3.x quarkus branch, the `quarkus-integration-test-reactive-messaging-pulsar` native integration test fails with:

```
========================================================================================================================
GraalVM Native Image: Generating 'quarkus-integration-test-reactive-messaging-pulsar-3.999-SNAPSHOT-runner' (executable)...
========================================================================================================================
For detailed information and explanations on the build output, visit:
https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md
------------------------------------------------------------------------------------------------------------------------
[1/8] Initializing...                                                                                   (12.0s @ 0.31GB)
 Java version: 25.0.4-beta+6-ea, vendor version: Mandrel-25.0.4.0-dev2c319ac4
 Graal compiler: optimization level: 2, target machine: x86-64-v3
 C compiler: gcc (linux, x86_64, 13.3.0)
 Garbage collector: Serial GC (max heap size: 80% of RAM)
 8 user-specific feature(s):
 - com.oracle.svm.thirdparty.gson.GsonFeature
 - io.quarkus.jackson.runtime.graal.JacksonSerializerRegistrationFeature: Conditionally registers Jackson SQL/XML serializers for reflection based on reachability
 - io.quarkus.runner.Feature: Auto-generated class by Quarkus from the existing extensions
 - io.quarkus.runtime.graal.DisableLoggingFeature: Adapts logging during the analysis phase
 - io.quarkus.runtime.graal.JVMChecksFeature
 - io.quarkus.runtime.graal.SkipConsoleServiceProvidersFeature: Skip unsupported console service providers when quarkus.native.auto-service-loader-registration is false
 - io.quarkus.security.BouncyCastleFeature
 - org.eclipse.angus.activation.nativeimage.AngusActivationFeature
------------------------------------------------------------------------------------------------------------------------
 5 experimental option(s) unlocked:
 - '-H:+AllowFoldMethods' (origin(s): command line)
 - '-H:BuildOutputJSONFile' (origin(s): command line)
 - '-H:-UseServiceLoaderFeature' (origin(s): command line)
 - '-H:+GenerateBuildArtifactsFile' (origin(s): command line)
 - '-H:AddOpens' (alternative API option(s): --add-opens java.base/java.lang=ALL-UNNAMED; origin(s): command line)
------------------------------------------------------------------------------------------------------------------------
Build resources:
 - 12.41GB of memory (74.0% of system memory, set via '-Xmx13g')
 - 4 thread(s) (100.0% of 4 available processor(s), determined at start)
[2/8] Performing analysis...  []                                                                         (2.7s @ 0.38GB)
    1,874 types,   2,306 fields, and     811 methods found reachable
    1,238 types,   2,175 fields, and  14,596 methods registered for reflection
        0 downcalls and 0 upcalls registered for foreign access

Error: com.oracle.svm.core.util.UserError$UserException: Class initialization of io.quarkus.runner.ApplicationImpl failed. Use the option 

    '--initialize-at-run-time=io.quarkus.runner.ApplicationImpl'

 to explicitly request initialization of this class at run time. Exception thrown by the class initializer:

java.lang.ClassNotFoundException: org.apache.pulsar.client.api.PulsarClientSharedResources
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageClassLoader.loadClass(NativeImageClassLoader.java:637)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:547)
	at io.smallrye.reactive.messaging.pulsar.PulsarConnector_Bean.<init>(Unknown Source)
	at io.quarkus.arc.setup.Default_ComponentsProvider_addBeans0.addBeans0(Unknown Source)
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:137)
	at io.quarkus.arc.Arc.initialize(Arc.java:53)
	at io.quarkus.arc.runtime.ArcRecorder.initContainer(ArcRecorder.java:53)
	at io.quarkus.runner.recorded.ArcProcessor$initializeContainer643029769.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.ArcProcessor$initializeContainer643029769.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1169)
	(internal stack frames of the image generator are omitted)


Caused by: com.oracle.svm.core.util.UserError$UserException: Class initialization of io.quarkus.runner.ApplicationImpl failed. Use the option 

    '--initialize-at-run-time=io.quarkus.runner.ApplicationImpl'

 to explicitly request initialization of this class at run time. Exception thrown by the class initializer:

java.lang.ClassNotFoundException: org.apache.pulsar.client.api.PulsarClientSharedResources
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageClassLoader.loadClass(NativeImageClassLoader.java:637)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:547)
	at io.smallrye.reactive.messaging.pulsar.PulsarConnector_Bean.<init>(Unknown Source)
	at io.quarkus.arc.setup.Default_ComponentsProvider_addBeans0.addBeans0(Unknown Source)
	at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Unknown Source)
	at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:137)
	at io.quarkus.arc.Arc.initialize(Arc.java:53)
	at io.quarkus.arc.runtime.ArcRecorder.initContainer(ArcRecorder.java:53)
	at io.quarkus.runner.recorded.ArcProcessor$initializeContainer643029769.deploy_0(Unknown Source)
	at io.quarkus.runner.recorded.ArcProcessor$initializeContainer643029769.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1169)
	(internal stack frames of the image generator are omitted)
```

See: https://github.com/graalvm/mandrel/actions/runs/28785799721/job/85353235986#step:12:4386

### Additional information

The smallrye messaging version bump from https://github.com/quarkusio/quarkus/pull/54794 is sufficient to reproduce the native integration test failure.

```
diff --git a/bom/application/pom.xml b/bom/application/pom.xml
index e92d08222b2..b3195f13ccb 100644
--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -58,7 +58,7 @@
         <smallrye-reactive-streams-operators.version>1.0.13</smallrye-reactive-streams-operators.version>
         <smallrye-reactive-types-converter.version>3.0.3</smallrye-reactive-types-converter.version>
         <smallrye-mutiny-vertx-binding.version>3.23.0</smallrye-mutiny-vertx-binding.version>
-        <smallrye-reactive-messaging.version>4.34.0</smallrye-reactive-messaging.version>
+        <smallrye-reactive-messaging.version>4.36.0</smallrye-reactive-messaging.version>
         <smallrye-stork.version>2.7.10</smallrye-stork.version>
         <jakarta.activation.version>2.1.4</jakarta.activation.version>
         <jakarta.annotation-api.version>3.0.0</jakarta.annotation-api.version>
```

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -212,7 +212,7 @@
         <sshd-common.version>2.17.1</sshd-common.version>
         <mime4j.version>0.8.12</mime4j.version>
         <mutiny-zero.version>1.2.1</mutiny-zero.version>
-        <pulsar-client.version>4.2.1</pulsar-client.version>
+        <pulsar-client.version>4.2.3</pulsar-client.version>
         <async-http-client.version>2.15.0</async-http-client.version>
         <!-- keep in-sync, if possible, with Micrometer registry Prometheus -->
         <prometheus.version>0.16.0</prometheus.version>

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -212,7 +212,7 @@
         <sshd-common.version>2.17.1</sshd-common.version>
         <mime4j.version>0.8.12</mime4j.version>
         <mutiny-zero.version>1.2.1</mutiny-zero.version>
-        <pulsar-client.version>3.3.0</pulsar-client.version>
+        <pulsar-client.version>4.2.1</pulsar-client.version>
         <async-http-client.version>2.15.0</async-http-client.version>
         <!-- keep in-sync, if possible, with Micrometer registry Prometheus -->
         <prometheus.version>0.16.0</prometheus.version>

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/DotNames.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/DotNames.java
@@ -50,6 +50,7 @@ final class DotNames {
     static final DotName PROTOBUF_GENERATED = DotName.createSimple("com.google.protobuf.GeneratedMessageV3");
     static final DotName PULSAR_SCHEMA = DotName.createSimple(org.apache.pulsar.client.api.Schema.class.getName());
     static final DotName PULSAR_AUTHENTICATION = DotName.createSimple(org.apache.pulsar.client.api.Authentication.class.getName());
+    static final DotName PULSAR_SSL_FACTORY = DotName.createSimple(org.apache.pulsar.common.util.PulsarSslFactory.class.getName());
 
     static final DotName LIST = DotName.createSimple(java.util.List.class.getName());
 

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
@@ -188,13 +188,13 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
         }
 
         Collection<ClassInfo> authPluginClasses = combinedIndex.getIndex()
-                .getAllKnownImplementations(DotNames.PULSAR_AUTHENTICATION);
+                .getAllKnownImplementors(DotNames.PULSAR_AUTHENTICATION);
         for (ClassInfo authPluginClass : authPluginClasses) {
             reflectiveClass.produce(ReflectiveClassBuildItem.builder(authPluginClass.name().toString())
                     .constructors().build());
         }
         Collection<ClassInfo> sslFactoryClasses = combinedIndex.getIndex()
-                .getAllKnownImplementations(DotNames.PULSAR_SSL_FACTORY);
+                .getAllKnownImplementors(DotNames.PULSAR_SSL_FACTORY);
         for (ClassInfo sslFactoryClass : sslFactoryClasses) {
             reflectiveClass.produce(ReflectiveClassBuildItem.builder(sslFactoryClass.name().toString())
                     .constructors().build());

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
@@ -164,6 +164,21 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
                 .builder("org.apache.pulsar.client.util.SecretsSerializer")
                 .constructors().build());
 
+        // Pulsar 4.2.x's DnsResolverGroupImpl / ConnectionPool use Netty's
+        // ReflectiveChannelFactory against the platform SocketChannel returned by
+        // EventLoopUtil. quarkus-netty registers only the NIO variants; register
+        // Epoll/KQueue/IOUring here so native image can find their no-arg
+        // constructors at runtime. Revisit when bumping pulsar-client — if Pulsar
+        // moves to non-reflective ChannelFactory factories, these can be removed.
+        reflectiveClass.produce(ReflectiveClassBuildItem
+                .builder("io.netty.channel.epoll.EpollSocketChannel",
+                        "io.netty.channel.epoll.EpollDatagramChannel",
+                        "io.netty.channel.kqueue.KQueueSocketChannel",
+                        "io.netty.channel.kqueue.KQueueDatagramChannel",
+                        "io.netty.incubator.channel.uring.IOUringSocketChannel",
+                        "io.netty.incubator.channel.uring.IOUringDatagramChannel")
+                .constructors().build());
+
         Collection<ClassInfo> authPluginClasses = combinedIndex.getIndex()
                 .getAllKnownImplementors(DotNames.PULSAR_AUTHENTICATION);
         for (ClassInfo authPluginClass : authPluginClasses) {

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
@@ -169,16 +169,23 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
         // ReflectiveChannelFactory against the platform SocketChannel returned by
         // EventLoopUtil. quarkus-netty registers only the NIO variants; register
         // Epoll/KQueue/IOUring here so native image can find their no-arg
-        // constructors at runtime. Revisit when bumping pulsar-client — if Pulsar
-        // moves to non-reflective ChannelFactory factories, these can be removed.
-        reflectiveClass.produce(ReflectiveClassBuildItem
-                .builder("io.netty.channel.epoll.EpollSocketChannel",
-                        "io.netty.channel.epoll.EpollDatagramChannel",
-                        "io.netty.channel.kqueue.KQueueSocketChannel",
-                        "io.netty.channel.kqueue.KQueueDatagramChannel",
-                        "io.netty.incubator.channel.uring.IOUringSocketChannel",
-                        "io.netty.incubator.channel.uring.IOUringDatagramChannel")
-                .constructors().build());
+        // constructors at runtime. These native transports are platform-specific
+        // and optional, so only register the ones actually on the classpath.
+        // Revisit when bumping pulsar-client — if Pulsar moves to non-reflective
+        // ChannelFactory factories, these can be removed.
+        String[] nettyTransportChannels = {
+                "io.netty.channel.epoll.EpollSocketChannel",
+                "io.netty.channel.epoll.EpollDatagramChannel",
+                "io.netty.channel.kqueue.KQueueSocketChannel",
+                "io.netty.channel.kqueue.KQueueDatagramChannel",
+                "io.netty.incubator.channel.uring.IOUringSocketChannel",
+                "io.netty.incubator.channel.uring.IOUringDatagramChannel"
+        };
+        for (String channel : nettyTransportChannels) {
+            if (QuarkusClassLoader.isClassPresentAtRuntime(channel)) {
+                reflectiveClass.produce(ReflectiveClassBuildItem.builder(channel).constructors().build());
+            }
+        }
 
         Collection<ClassInfo> authPluginClasses = combinedIndex.getIndex()
                 .getAllKnownImplementations(DotNames.PULSAR_AUTHENTICATION);

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
@@ -197,6 +197,14 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
                 .addNativeImageSystemProperty("io.netty.handler.ssl.noOpenSsl", "true")
                 .addRuntimeInitializedClass("org.apache.pulsar.common.allocator.PulsarByteBufAllocator")
                 .addRuntimeInitializedClass("org.apache.pulsar.common.protocol.Commands")
+                // Pulsar 4.2.x's LightProtoCodec computes Unsafe field offsets
+                // (String.value field offset, byte[] base offset) in its static
+                // initializer. GraalVM cannot auto-recompute these, so building
+                // them at image-build time bakes in host-JVM offsets and segfaults
+                // at runtime in writeRawString. Initialize at run time so the
+                // offsets match the SubstrateVM heap layout. Revisit when bumping
+                // pulsar-client — Pulsar 25883 embeds this config upstream.
+                .addRuntimeInitializedClass("org.apache.pulsar.common.api.proto.LightProtoCodec")
                 .addRuntimeInitializedClass("org.apache.pulsar.client.impl.auth.oauth2.protocol.TokenClient")
                 .addRuntimeInitializedClass("org.apache.pulsar.client.impl.crypto.MessageCryptoBc")
                 .addRuntimeInitializedClass("org.apache.pulsar.client.impl.schema.generic.GenericProtobufNativeSchema")

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
@@ -131,8 +131,9 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
     }
 
     @BuildStep
-    IndexDependencyBuildItem indexPulsar() {
-        return new IndexDependencyBuildItem("org.apache.pulsar", "pulsar-client-original");
+    void indexPulsar(BuildProducer<IndexDependencyBuildItem> indexDependency) {
+        indexDependency.produce(new IndexDependencyBuildItem("org.apache.pulsar", "pulsar-client-original"));
+        indexDependency.produce(new IndexDependencyBuildItem("org.apache.pulsar", "pulsar-common"));
     }
 
     @BuildStep
@@ -180,9 +181,15 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
                 .constructors().build());
 
         Collection<ClassInfo> authPluginClasses = combinedIndex.getIndex()
-                .getAllKnownImplementors(DotNames.PULSAR_AUTHENTICATION);
+                .getAllKnownImplementations(DotNames.PULSAR_AUTHENTICATION);
         for (ClassInfo authPluginClass : authPluginClasses) {
             reflectiveClass.produce(ReflectiveClassBuildItem.builder(authPluginClass.name().toString())
+                    .constructors().build());
+        }
+        Collection<ClassInfo> sslFactoryClasses = combinedIndex.getIndex()
+                .getAllKnownImplementations(DotNames.PULSAR_SSL_FACTORY);
+        for (ClassInfo sslFactoryClass : sslFactoryClasses) {
+            reflectiveClass.produce(ReflectiveClassBuildItem.builder(sslFactoryClass.name().toString())
                     .constructors().build());
         }
 

--- a/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
+++ b/extensions/smallrye-reactive-messaging-pulsar/deployment/src/main/java/io/quarkus/smallrye/reactivemessaging/pulsar/deployment/SmallRyeReactiveMessagingPulsarProcessor.java
@@ -181,7 +181,6 @@ public class SmallRyeReactiveMessagingPulsarProcessor {
                 .addRuntimeInitializedClass("org.apache.pulsar.client.impl.ConnectionPool")
                 .addRuntimeInitializedClass("org.apache.pulsar.client.impl.ControlledClusterFailover")
                 .addRuntimeInitializedClass("org.apache.pulsar.client.impl.HttpClient")
-                .addRuntimeInitializedClass("org.apache.pulsar.client.util.WithSNISslEngineFactory")
                 .addRuntimeInitializedClass("com.yahoo.sketches.quantiles.DoublesSketch")
                 .addRuntimeInitializedClass("io.netty.buffer.PooledByteBufAllocator")
                 .addRuntimeInitializedClass("io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledHeapByteBuf")

